great_expectations/core/partitioners.py

```diff
@@ -0,0 +1,64 @@
+from typing import List, Literal, Union
+
+from great_expectations.compatibility import pydantic
+
+
+class PartitionerYear(pydantic.BaseModel):
+    column_name: str
+    method_name: Literal["partition_on_year"] = "partition_on_year"
+
+
+class PartitionerYearAndMonth(pydantic.BaseModel):
+    column_name: str
+    method_name: Literal["partition_on_year_and_month"] = "partition_on_year_and_month"
+
+
+class PartitionerYearAndMonthAndDay(pydantic.BaseModel):
+    column_name: str
+    method_name: Literal[
+        "partition_on_year_and_month_and_day"
+    ] = "partition_on_year_and_month_and_day"
+
+
+class PartitionerDatetimePart(pydantic.BaseModel):
+    datetime_parts: List[str]
+    column_name: str
+    method_name: Literal["partition_on_date_parts"] = "partition_on_date_parts"
+
+
+class PartitionerDividedInteger(pydantic.BaseModel):
+    divisor: int
+    column_name: str
+    method_name: Literal[
+        "partition_on_divided_integer"
+    ] = "partition_on_divided_integer"
+
+
+class PartitionerModInteger(pydantic.BaseModel):
+    mod: int
+    column_name: str
+    method_name: Literal["partition_on_mod_integer"] = "partition_on_mod_integer"
+
+
+class PartitionerColumnValue(pydantic.BaseModel):
+    column_name: str
+    method_name: Literal["partition_on_column_value"] = "partition_on_column_value"
+
+
+class PartitionerMultiColumnValue(pydantic.BaseModel):
+    column_names: List[str]
+    method_name: Literal[
+        "partition_on_multi_column_values"
+    ] = "partition_on_multi_column_values"
+
+
+Partitioner = Union[
```

Comment:
Very non-blocking, and I think we discussed this synchronously, but is there a benefit of building the union vs inheriting from a `Partitioner` ABC? If we went the ABC route, and moved to using enums, we could potentially increase our type safety with something like.

```
_T = TypeVar("_T", bound=MethodName)

class Partitioner(Generic[_T], ABC, pydantic.BaseModel):
  method_name: _T          # though IIRC, we'd still have to set the default in the child class.
```

Reply:
thanks for these suggestions - looks like both could potentially be implemented without changing behavior outside of the `partitioners` module, so i'll look into them as a followup PR 👍